[EOSF-770] Add file-editor widget

baylee-d opened the pull request:
## Ticket

https://openscience.atlassian.net/browse/EOSF-770

# Purpose

Add a file-edit widget to the ember-osf for use on the file-detail page in quickfiles.

# Summary of changes

- Add a file-edit widget that uses ember-ace to match the OSF file editor
- Create a new const file that has all the mime-types for editable files

# Screenshots

File editor at half screen (shares the screen with the file viewer)
![screen shot 2017-10-09 at 1 48 35 pm](https://user-images.githubusercontent.com/19379783/31351556-8a960202-acf9-11e7-8534-5437f4047d4d.png)

File editor at full screen
![screen shot 2017-10-09 at 1 48 48 pm](https://user-images.githubusercontent.com/19379783/31351575-9b981ee6-acf9-11e7-91e4-7a93552e23d6.png)

hmoco commented:
Doing functional review:

hmoco commented:
There should be a "saved" indicator when pressing the save button and action is successful, and error handling if it isn't. Right now if it fails an error is sent to the console, but nothing to the user.

hmoco commented:
osf.io's editor seems to have syntax highlighting (probably an ace editor option). Can we turn this on  here?

hmoco commented:
Looks great otherwise! Moving to code review!

hmoco commented on addon/const/mime-types.js:
```diff
@@ -0,0 +1,449 @@
+
+/**
+ * Provides list of all file extensions that can be
+ * rendered as text by MFR
```
If this list was retrieved from somewhere else, would be good to provide a link in the docstring here.

baylee-d replied:
I got this list from osf.io (https://github.com/baylee-d/osf.io/blob/develop/website/static/js/mime.js).  Should there still be a link?

hmoco replied:
Yeah, but point to this: https://github.com/CenterForOpenScience/osf.io/blob/develop/website/static/js/mime.js

Like I've done here maybe: https://github.com/CenterForOpenScience/ember-osf/pull/240/files#diff-26748c589710c70e6bfe023bc8a3c072R21

Not sure what's the standard way of doing this, but I think if we can't directly import it, we should attribute it

baylee-d replied:
Ah, okay that makes sense 👍

hmoco commented on tests/integration/components/file-editor/component-test.js:
```diff
@@ -0,0 +1,37 @@
+    const editor = ace.edit(document.querySelector('[data-ember-ace]'));
+    assert.equal(editor.getSession().getValue(), 'Test text');
+
+});
```
Is it possible to add a test checking the reverse functionality?

baylee-d commented:
- Changed the test for revert functionality and added a link to the proper file in osf.io for the mime-types.  

- Toastr notifications have been added on the ember-osf-web side for on save/fail.  

It was also decided that syntax highlighting can be done in a later iteration in favor of getting this out.

jamescdavis commented on addon/components/file-editor/component.js:
```diff
@@ -0,0 +1,36 @@
+        },
+        saveText() {
+            const editor = ace.edit(document.querySelector('[data-ember-ace]'));
+            this.sendAction('save', editor.getSession().getValue());
```
Please use a closure action.

jamescdavis commented on addon/components/file-editor/component.js:
```diff
@@ -0,0 +1,36 @@
+
+    actions: {
+        revertText() {
+            const editor = ace.edit(document.querySelector('[data-ember-ace]'));
```
Instead of pulling this out of the DOM, should use the [`ready` action](https://github.com/dfreeman/ember-ace#core) to set the editor.

jamescdavis commented on addon/components/file-editor/component.js:
```diff
@@ -0,0 +1,36 @@
+            editor.getSession().setValue(this.get('fileText'));
+        },
+        saveText() {
+            const editor = ace.edit(document.querySelector('[data-ember-ace]'));
```
SAA

jamescdavis commented on addon/components/file-editor/component.js:
```diff
@@ -0,0 +1,43 @@
+        revertText() {
+            const fileText = this.get('fileText');
+            this.set('fileText', '');
+            Ember.run.next(() => this.set('fileText', fileText));
```
Can you add a comment about why this is done in run.next?

jamescdavis commented on addon/components/file-editor/style.scss:
```diff
@@ -0,0 +1,17 @@
+    width: 100%;
+    border: 1px solid lightgray;
+    cursor: text;
+}
```
We should put these styles in a parent class for the component.

jamescdavis commented on addon/components/file-editor/template.hbs:
```diff
@@ -0,0 +1,28 @@
+</div>
+<div>
+    <div class='editor-pane panel-body'>
+        <div style='padding-top:10px;'>
```
Should put this in the .scss

jamescdavis commented on tests/integration/components/file-editor/component-test.js:
```diff
@@ -0,0 +1,67 @@
+    this.render(hbs`
+        {{file-editor
+            fileText=''
+            save='save'
```
I don't think you need this unless you're testing the save button.

jamescdavis commented on addon/components/file-editor/style.scss:
```diff
@@ -0,0 +1,20 @@
+}
+.file-editor .text-editor-container {
+    padding-top: 10px;
+}
```
in SASS you can do:
```
.file-editor {
    .panel-heading {
        text-align: left;
        .panel-title {
            font-weight: 400;
        }
    }
    .text-editor-container {
         padding-top: 10px;
    }
}
```

jamescdavis commented on tests/integration/components/file-editor/component-test.js:
```diff
@@ -0,0 +1,81 @@
+
+    this.set('externalSaveAction', (actual) => {
+        let expected = 'Test to save this new text!';
+        assert.deepEqual(actual, expected, 'Save function properly passes new value');
```
don't really need to use `deepEqual` here since it's just a string

jamescdavis commented on tests/integration/components/file-editor/component-test.js:
```diff
@@ -0,0 +1,81 @@
+    // Tests that the save function works
+
+    this.set('externalSaveAction', (actual) => {
+        let expected = 'Test to save this new text!';
```
const?
